### PR TITLE
Logging implementation for tendrl

### PR DESCRIPTION
**Pull request by GowthamShanmugam**
tendrl-bug-id: Tendrl/specifications#28
tendrl-spec: logging_implementation

Signed-off-by: root <root@dhcp42-75.lab.eng.blr.redhat.com>

### CODE REVIEW

**Comment by GowthamShanmugam** [comment]
@r0h4n review please

**Comment by GowthamShanmugam** [comment]
@nthomas-redhat @anmolbabu @brainfunked @anmolsachan @shtripat  please review this

File: etc/tendrl/node-agent/node-agent.conf.yaml.sample [hunk 1]
```diff
@@ -5,3 +5,4 @@ tendrl_ansible_exec_file: $HOME/.tendrl/node-agent/ansible_exec
 log_cfg_path: /etc/tendrl/node-agent/node-agent_logging.yaml
 log_level: DEBUG
 package_source_type: rpm
+logging_socket_path: "/tmp/node-agent/.node_agent.sock"
```
Comment (nnDarshan) [inline etc/tendrl/node-agent/node-agent.conf.yaml.sample]: I feel its better to have the socket file at "/var/run/tendrl/node-agent/.node_agent.sock" as /tmp is a bit vulnerable, user might clear /tmp directory.
@nthomas-redhat @shtripat please comment

File: tendrl/node_agent/message/handler.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            self.bind_unix_listener(),
+            self.read_socket
+        )
+
+    def read_socket(self, sock, address):
+        try:
+            data = sock.recv(RECEIVE_DATA_SIZE)
+            message = Message.from_json(data)
+            Logger(message)
+        except (TypeError, ValueError, KeyError,
```
Comment (nnDarshan) [inline tendrl/node_agent/message/handler.py]: whats the point of these try except blocks ? if you are just raising the exception. This comment holds good for all the below try except blocls

File: tendrl/node_agent/message/handler.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+                socket_error, socket_timeout) as ex:
+            raise ex
+
+    def _run(self):
+        try:
+            self.server.serve_forever()
+        except (TypeError, BlockingIOError, socket_error, ValueError) as ex:
+            raise ex
+
+    def stop(self):
```
Comment (nnDarshan) [inline tendrl/node_agent/message/handler.py]: Does this take care of deleting the socket file ? If no, it has to be taken care. Else user will have to manually delete the socket file after every restart

File: tendrl/node_agent/message/handler.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+from gevent.server import StreamServer
+from io import BlockingIOError
+import os
+from tendrl.commons.message import Message
+from tendrl.node_agent.message.logger import Logger
+
+
+RECEIVE_DATA_SIZE = 4096
+
+class MessageHandler(gevent.greenlet.Greenlet):
```
Comment (nnDarshan) [inline tendrl/node_agent/message/handler.py]: If any issue occurs in this class, I feel it has to be written to stdout as the logger is not initialized yet and if not written to any place there is no way for user to know what went wrong.

File: etc/tendrl/node-agent/node-agent.conf.yaml.sample (hunk 1 above)
Comment (GowthamShanmugam) [inline etc/tendrl/node-agent/node-agent.conf.yaml.sample]: @r0h4n please give your suggestion for this

File: tendrl/node_agent/message/handler.py (hunk 4 above)
Comment (GowthamShanmugam) [inline tendrl/node_agent/message/handler.py]: @nnDarshan if the user runs node_agent as service then how they will know where exception occurred, this problem also in event.py in commons, if socket fails i don't know what to do, if socket fails i am raising an exception in event.py but try catch is required for all the places where log message is created. we can handle error there but problem is for alerting anmol wants that exception in collecd.

File: tendrl/node_agent/message/handler.py (hunk 2 above)
Comment (GowthamShanmugam) [inline tendrl/node_agent/message/handler.py]: @nnDarshan should we print the exception?

File: tendrl/node_agent/message/handler.py (hunk 3 above)
Comment (GowthamShanmugam) [inline tendrl/node_agent/message/handler.py]: @nnDarshan i will do it

File: tendrl/node_agent/message/logger.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            else:
+                tendrl_ns.etcd_orm.client.write(
+                    ("nodes/%s/Messages/%s") % (
+                        self.message.node_id,
+                        self.message.message_id),
+                    Message.to_json(self.message)
+                )
+            # storing messages global wise
+            tendrl_ns.etcd_orm.client.write(
+                ("Messages/%s") % (
```
Comment (anmolbabu) [inline tendrl/node_agent/message/logger.py]: Should this be lower case just as per normal convention "Messages" ==> "messages" ?

File: tendrl/node_agent/message/logger.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+                    self.message.message_id),
+                Message.to_json(self.message)
```
Comment (anmolbabu) [inline tendrl/node_agent/message/logger.py]: :+1: for dumping as json to etcd HERE

File: tendrl/node_agent/message/logger.py (hunk 5 above)
Comment (GowthamShanmugam) [inline tendrl/node_agent/message/logger.py]: @anmolbabu all the place we are using caps only, rohan suggested to me to like this

File: etc/tendrl/node-agent/node-agent.conf.yaml.sample (hunk 1 above)
Comment (brainfunked) [inline etc/tendrl/node-agent/node-agent.conf.yaml.sample]: `/var/run/tendrl/message.sock`.

**Comment by GowthamShanmugam** [comment]
@brainfunked i will change it

**Comment by GowthamShanmugam** [comment]
@nthomas-redhat @brainfunked @shtripat @nnDarshan @anmolbabu @r0h4n @TimothyAsir @anmolsachan  please review

File: tendrl/node_agent/message/handler.py (hunk 4 above)
Comment (nthomas-redhat) [inline tendrl/node_agent/message/handler.py]: see the comment-->https://github.com/Tendrl/specifications/pull/94#issuecomment-279876382

File: tendrl/node_agent/message/handler.py (hunk 2 above)
Comment (nthomas-redhat) [inline tendrl/node_agent/message/handler.py]: Exceptions should always go into stderr along with configured logger

File: tendrl/node_agent/message/handler.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            print(traceback.format_exc())
+
+    def _run(self):
+        try:
+            self.server.serve_forever()
+        except (TypeError, BlockingIOError, socket_error, ValueError) as ex:
+            print(traceback.format_exc())
+
+    def stop(self):
+        self.sock.close()
```
Comment (nthomas-redhat) [inline tendrl/node_agent/message/handler.py]: Also handle the exceptions in this function

File: tendrl/node_agent/message/logger.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
+import logging
+from tendrl.commons.message import Message
+
+
+LOG = logging.getLogger(__name__)
+
+
+class Logger(object):
+    logger_priorities = {"notice": "info",
+                         "info": "info",
```
Comment (nthomas-redhat) [inline tendrl/node_agent/message/logger.py]: Looks like 'info' and 'notice' are same. why you need both?
Is it done for compatebility with common logging?

Comment (GowthamShanmugam) [inline tendrl/node_agent/message/logger.py]: @nthomas-redhat alert are always comes as notice, anmol can pick the alert messages from etcd which is in notice type. 

We are storing all the messages in  etcd under /Messages. Anmol can tack alert from etcd using which have type notice.  But in log there is no notice function so i logging notice as info

**Comment by GowthamShanmugam** [comment]
@nthomas-redhat @shtripat @nnDarshan @r0h4n @brainfunked @anmolbabu @TimothyAsir please review

File: tendrl/node_agent/message/handler.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+from gevent.socket import error as socket_error
+from gevent.socket import timeout as socket_timeout
+from io import BlockingIOError
+import os
+import sys
+from tendrl.commons.message import Message
+from tendrl.node_agent.message.logger import Logger
+
+
+RECEIVE_DATA_SIZE = 4096
```
Comment (brainfunked) [inline tendrl/node_agent/message/handler.py]: What's the rationale behind this specific size?

Comment (GowthamShanmugam) [inline tendrl/node_agent/message/handler.py]: @brainfunked When reading from a socket it's always required to specify reading byte size, it means never return until read x bytes from the socket.

**Pull request closed without merging**